### ISSUE TITLE
Report allocation errors through the panic handler

### ISSUE THREAD

**Issue by Amanieu** [rust-lang/libs-team#192]
# Proposal

## Problem statement

We currently use a completely separate mechanism to handle allocation errors (OOM) than panics, despite both of these being very similar. However both are very similar: they involve handling runtime failures that are generally unrecoverable. It would be better to unify them under a single mechanism which is already stable: panic handlers (no_std) and panic hooks (std).

## Motivation, use-cases

This change has already been done for `no_std` in https://github.com/rust-lang/rust/issues/66741: it was noticed that the vast majority of users of `#[alloc_error_handler]` were just calling panic anyways. This avoided the need to stabilize the `alloc_error_handler` attribute by defaulting to a panic if one was not specified.

The OOM hook API has also remained unstable for a very long time with no clear path to stabilization. The best way forward would be to unify it with the panic hook API just like it has been done in no_std.

Some use cases ([firefox](https://github.com/rust-lang/rust/issues/51540#issuecomment-1452662139)) still need access to the size of the failed allocation, so this is provided as a custom panic payload.

## Solution sketches

```
// alloc::alloc

pub struct AllocErrorPanicPayload { .. }

impl AllocErrorPanicPayload {
    pub fn layout(&self) -> Layout;
}
```

This ACP proposes to make the following changes:
- Introduce the `AllocErrorPanicPayload` type which holds the `Layout` of a failed allocation.
- Make `handle_alloc_error` call the normal panic handler with a special payload of `AllocErrorPanicPayload`.
  - This can be observed by panic hooks with std or by the panic handler with no_std.
  - Unless `-Zoom=panic` is used, this is a [non-unwinding panic](https://doc.rust-lang.org/nightly/core/panic/struct.PanicInfo.html#method.can_unwind): if the panic hook returns then the process is aborted. This maintains the current behavior of aborting on OOM by default.
- Remove the unstable `alloc_error_hook` std API (https://github.com/rust-lang/rust/issues/51245). Its job is now handled by panic hooks.
- Remove the unstable `#[alloc_error_handler]` attribute (https://github.com/rust-lang/rust/issues/51540) for no_std programs. Its job is now handled by `#[panic_handler]`.

In the standard library, no additional allocations are made as part of the OOM handling process, except in 1 specific situation:
- With `-Zoom=panic`, if the panic hook returns then an allocation is necessary to box the payload and allocate an exception object. However it's not a big deal if this, fails, the process will simply abort immediately in that case.

## Links and related work

Tracking issues that will be closed by this change:
- https://github.com/rust-lang/rust/issues/51245 `alloc_error_hook`
- https://github.com/rust-lang/rust/issues/51540 `alloc_error_handler`
  - [Discussion about removing `alloc_error_handler`](https://github.com/rust-lang/rust/issues/51540#issuecomment-1448404145)

## What happens now?

This issue is part of the libs-api team [API change proposal process]. Once this issue is filed the libs-api team will review open proposals in its weekly meeting. You should receive feedback within a week or two.

[API change proposal process]: https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html

**Comment by Amanieu** [comment]
I don't see this as a strong argument to avoid using payloads, it's just that payloads are underused today. Treating OOM as a panic is almost always what people want, and allows the resulting error message to be properly integrated into whatever error reporting mechanism an application is using (e.g. uploading crash reports).

**Comment by m-ou-se** [comment]
Using payloads in a std panic hook is fine. I'm saying that using payloads in a #[panic_handler] would be a entirely new concept. It's true that we already have a .payload() method available in #[panic_handler]s, but it doesn't do anything.

A (std) panic hook is expected to do `.payload().downcast_ref<String>()` (and &str) to get the panic message, but that will always return `None` in a `#[panic_handler]`. So, we can never get the concept of a 'payload' to mean the same thing in panic handlers and panic hooks. So I disagree this is a 'unified single mechanism'.

It's a valid discussion to have whether we should start exposing the concept of a panic payload in #[panic_handler], but that would be an entirely new thing, not an existing (unified) mechanism.

(And on that discussion: I don't think we should start using payloads in #[panic_handler], because it will never mean the same thing as a payload in a panic hook: it will not a payload for formatted panic messages. And `std` will need to take ownership of a panic payload to be able to unwind/catch, which isn't possible when just borrowing the payload like a #[panic_handler] does.)

**Comment by m-ou-se** [comment]
> Treating OOM as a panic is almost always what people want, and allows the resulting error message to be properly integrated into whatever error reporting mechanism an application is using (e.g. uploading crash reports).

That will happen regardless of whether we have a AllocErrorPanicPayload payload mechanism or a default alloc_error_handler that panics, right?

**Comment by RalfJung** [comment]
> Using payloads in a std panic hook is fine. I'm saying that using payloads in a #[panic_handler] would be a entirely new concept. It's true that we already have a .payload() method available in #[panic_handler]s, but it doesn't do anything.

I've been wondering before why panic hooks and panic handlers take the same type as argument. In a sense there's really two different types here that are just merged into one -- "PanicInfo with a message" and "PanicInfo with a payload".

**Comment by RalfJung** [comment]
Yeah, that resolves the concern you raised [above](https://github.com/rust-lang/libs-team/issues/192#issuecomment-1743062566), I assume. Or at least the concern would have to be phrased differently now I guess?

It doesn't resolve my concern, though.

**Comment by m-ou-se** [comment]
It "resolves" all concerns because it means this acp is no longer possible.

**Comment by RalfJung** [comment]
Custom panic payloads are still a thing though, aren't they?

**Comment by m-ou-se** [comment]
Only in std. [`std::panic::panic_any`](https://doc.rust-lang.org/stable/std/panic/fn.panic_any.html) will directly call the hook you set with [`std::panic::set_hook`](https://doc.rust-lang.org/stable/std/panic/fn.set_hook.html) with any payload, but that doesn't go through any of (core's) panic machinery.

**Comment by RalfJung** [comment]
Ah I see, and this here is meant to work for `alloc` even without `std`.

So... is there still a plan for what to do with alloc error handlers?

**Comment by m-ou-se** [comment]
I think Externally Implementable Items (EII) are the solution.

@jdonszelmann and I are close to getting that implemented. (There's a working implementation that only needs some debugging and cleaning up.)

It will allow `alloc` to do:

```rust
#[externally_implementable]
fn alloc_error_handler(_: Layout) -> ! {
    panic!("default message");
}
```

And then a user can override it with something like:

```rust
#[alloc::alloc_error_handler]
fn my_alloc_error_handler(layout: Layout) -> ! {
    // ...
}
```

**Comment by m-ou-se** [comment]
If we do that, then what we do in std is an open question.

Possible answers:

1. Do nothing in std. The user can just override the `#[alloc_error_handler]` if they want.
2. Implement the `#[alloc_error_handler]` in std (just like how it implements the `#[panic_handler]`), and call the panic hook from it with a special `AllocErrorPanicPayload` payload.

Option 2 is basically doing this ACP for std only.

One could argue that option 2 is unnecessary if option 1 already works. But it's also arguable that with `std` it's more consistent to have everything end up in the (dynamically/runtime settable) panic hook.

**Comment by RalfJung** [comment]
That sounds like it would allow safe code to run arbitrary code on allocation failure, which is a big soundness footgun as described [above](https://github.com/rust-lang/libs-team/issues/192#issuecomment-2790203837).

**Comment by m-ou-se** [comment]
Yeah, we support unsafe EIIs. We can mark the EII as unsafe with something like:

```rust
#[externally_implementable(unsafe)]
fn alloc_error_handler(_: Layout) -> ! {
    abort();
}
```

And then a user needs to use an unsafe attribute to override it:

```
#[unsafe(alloc::alloc_error_handler)]
fn my_alloc_error_handler(layout: Layout) -> ! {
    // ...
}
```

**Comment by m-ou-se** [comment]
But it sounds like option 2 isn't possible then, because `set_hook` is safe.
